client/client.go

```diff
@@ -300,43 +303,66 @@ func (client *Client) Hijack(f func(*http.Request) (*http.Response, error)) {
 	client.doer = hijacked{f}
 }
 
-type doFlags struct {
-	NoTimeout bool
+type doOptions struct {
+	// Timeout is the overall request timeout
+	Timeout time.Duration
+	// Retry interval
+	Retry time.Duration
+}
+
+func ensureDoOpts(opts *doOptions) *doOptions {
+	if opts == nil {
+		// defaults
+		opts = &doOptions{
+			Timeout: doTimeout,
+			Retry:   doRetry,
+		}
+	}
+	return opts
+}
+
+// doNoTimeoutAndRetry can be passed to the do family to not have timeout
+// nor retries.
+var doNoTimeoutAndRetry = &doOptions{
+	Timeout: time.Duration(-1),
 }
 
 // do performs a request and decodes the resulting json into the given
 // value. It's low-level, for testing/experimenting only; you should
 // usually use a higher level interface that builds on this.
-func (client *Client) do(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}, flags doFlags) (statusCode int, err error) {
-	retry := time.NewTicker(doRetry)
-	defer retry.Stop()
-	timeout := time.NewTimer(doTimeout)
-	defer timeout.Stop()
+func (client *Client) do(method, path string, query url.Values, headers map[string]string, body io.Reader, v interface{}, opts *doOptions) (statusCode int, err error) {
+	opts = ensureDoOpts(opts)
 
 	var rsp *http.Response
 	var ctx context.Context = context.Background()
-	for {
-		if flags.NoTimeout {
-			rsp, err = client.raw(ctx, method, path, query, headers, body)
-		} else {
+	if opts.Timeout < 0 {
+		// no timeout and retries
+		rsp, err = client.raw(ctx, method, path, query, headers, body)
+	} else {
+		retry := time.NewTicker(opts.Retry)
```

Comment:
NewTicker will panic if opts.Retry is less than or equal to zero, so we should verify this is greater than or equal to zero too

Reply:
did this